### PR TITLE
Use dojo-core/load to resolve modules

### LINKED ISSUE #68: Utilise dojo-core/load for dynamic loading of modules
**Enhancement**

We should consider using `dojo-core/load` to dynamically load modules, which hopefully we can continue to abstract away any logic that would become problematic when using a non-AMD loader and any dynamic loading features of `app`.

This would result in us contracting that `load` would be able to continue to load certain resources that are then loaded into a particular context (e.g. executable code) and `dojo-core/request` would focus on retrieving resources which are generally not executed/evaluated (e.g. CSS, HTML, etc...)

### PR DESCRIPTION
**Pull request by novemberborn**
Fixes #68.

### CODE REVIEW

**Comment by novemberborn** [comment]
I suspect CI might fail due to <https://github.com/dojo/core/issues/209>. Regardless we shouldn't land before that's fixed.

File: src/lib/moduleResolver.ts [hunk 1]
```diff
@@ -38,26 +39,23 @@ export const RESOLVE_CONTENTS = Symbol();
...
-				return module.__esModule ? module.default : module;
-			}
-			else if (member === RESOLVE_CONTENTS) {
-				const contents: { [member: string]: any } = {};
-				for (const member of Object.keys(module)) {
-					if (member !== '__esModule' && member !== 'default') {
-						contents[member] = module[member];
+		return load(toAbsMid(mid))
+			.then(([ module ]) => {
+				if (member === 'default') {
```
Comment (kitsonk) [inline src/lib/moduleResolver.ts]: Should we eventually bake this into `load`?

**Comment by novemberborn** [comment]
Need to investigate how this interacts with `ToAbsMid`, and whether that should still be supported when it comes to Webpack compilation.

**Comment by kitsonk** [comment]
I think we discussed offline that it would be best to bake that into `load` too... so `load()` can deal with relative or absolute MIDs.  The only thing that would need to be passed then is some sort of context though.  The "rule" we had in Dojo 1 parser, IIRC, was that relative MIDs were disallowed in markup though, because "relative to what" was often just too confusing to reason out.

**Comment by novemberborn** [comment]
> I think we discussed offline that it would be best to bake that into load too... so load() can deal with relative or absolute MIDs.

Yup, that's fine.

> The "rule" we had in Dojo 1 parser, IIRC, was that relative MIDs were disallowed in markup though, because "relative to what" was often just too confusing to reason out.

Agreed, we can do that.

> The only thing that would need to be passed then is some sort of context though. 

That's the tricky bit. It seems to me that even with POJO definitions it's hard to reason about what they're relative to. So either we disallow relative MIDs entirely, or we only allow them if the app was created with a base. But then they'd be allowed in markup and in POJO definitions.

**Comment by kitsonk** [comment]
A sufficient fix is in `dojo-core` master, but now needs a release.

**Comment by novemberborn** [comment]
> A sufficient fix is in dojo-core master, but now needs a release.

For using `dojo-core/load`, yes. What should we do with relative MIDs?

> So either we disallow relative MIDs entirely, or we only allow them if the app was created with a base. But then they'd be allowed in markup and in POJO definitions.

**Comment by kitsonk** [comment]
I would say disallow relative MIDs. @matt-gadd thoughts?

**Comment by matt-gadd** [comment]
Really for Webpack relative paths would actually be preferred, but I know we're going to run into inconsistencies with that and an AMD relative path. I guess the simplest solution for now would be to disallow them.

**Comment by kitsonk** [comment]
Well, to pontificate, it would be a heck of a lot easier to create some sort of MID cache in load in the future with a full MID than try to continue to support relative MIDs... It is always hard in declarative markup to agree where it is relative to.  For example, is it the HTML file?  What if the HTML is simply fulfilled via something like request, or is set as a property of something else which in turn fetches it... It is just madness that we could add back in in the future if there was a concrete use case.

**Comment by matt-gadd** [comment]
@kitsonk agreed.

**Comment by dylans** [comment]
@rorticus this is stale after some changes to @dojo/core.load... perhaps you can take a look and refactor and land this.

**Comment by matt-gadd** [comment]
@dylans I think we should close this, it's likely that this will no longer be needed as you can dynamically load widgets via `widget-core` factory registries now.

**Pull request closed without merging**